File: Packages/TypeEngine/Sources/TypeEngine/PerTapCostProvider.swift
import Foundation

/// One captured keystroke touch point (PLAN.md "Touch decoding", stage 1).
///
/// `dxNorm`/`dyNorm` are within-key normalized offsets from the RESOLVED
/// key's center: −0.5…+0.5 at the key's touch-cell edges (the cell spans
/// the full key pitch, visual insets included), x growing rightward and y
/// growing DOWNWARD (toward the spacebar) — the same convention as the
/// ReplayRig TSI traces (`tsi-distributions.json`) and
/// `Learning.LearningEvent.touchSample`.
///
/// `char` is the resolved character, lowercased by the producer
/// (`TypingSession.noteTap`).
public struct TapSample: Equatable, Sendable {
    public let char: Character
    public let dxNorm: Double
    public let dyNorm: Double

    public init(char: Character, dxNorm: Double, dyNorm: Double) {
        self.char = char
        self.dxNorm = dxNorm
        self.dyNorm = dyNorm
    }
}

/// The coordinate-evidence `PositionCostProvider` (PLAN.md "Touch
/// decoding", stage 1 — the provider the `BeamDecoder` seam was built for).
///
/// Substitution pricing: a 2D Gaussian at the ACTUAL tap point. STAGE 1
/// (always available): an axis-aligned Gaussian at every key's center, σ
/// seeded from the Google TSI within-key distributions
/// (`EngineConfig.tapSigmaX`/`tapSigmaY`; tunable). Costs are LOG-LIKELIHOOD
/// RATIOS against the resolved key, so the protocol's "0 when equal"
/// convention holds and the price works BOTH ways (the bidirectional
/// evidence principle):
///
///  * a tap near the boundary toward a neighbor makes that neighbor's
///    substitution cheap — down to the `minSubstitution` floor, well below
///    the ~1-nat static key-distance price (corrections get ENABLED),
///  * a dead-center tap prices every neighbor substitution near the
///    Gaussian exponent of a full key pitch (≈ 7 nats at the TSI σx),
///    capped at `maxSubstitution` — the correction VETO: believe the user.
///
/// STAGE 2 (this wave — the per-user adaptive touch model): when a
/// `PersonalTouchSnapshot` is injected, every key whose effective sample
/// count clears `EngineConfig.touchPersonalMinSamples` is priced against
/// its PERSONAL Gaussian — mean-shifted center, per-key covariance —
/// shrunk toward the TSI prior by w = n/(n + touchPriorStrength):
///
///   mean_blend = w·mean_personal        var_blend = w·var + (1−w)·σ_TSI²
///   cov_blend  = w·cov_personal
///
/// Because per-key Gaussians now differ, the ½·ln|Σ| normalizer no longer
/// cancels between keys; it is carried per key RELATIVE to the prior's
/// (so prior-priced keys keep the exact stage-1 numbers). Numerical
/// safety, per key: blended σ floored at `touchSigmaFloor`, correlation
/// clamped to |ρ| ≤ 0.9, and any non-finite/degenerate aggregate drops the
/// key back to the static prior — a corrupt personal model can never
/// poison pricing, and every priced cost still passes the existing
/// `minSubstitution`/`maxSubstitution` clamps. With no snapshot, or no key
/// past the gate, the arithmetic is the stage-1 code path, byte-identical.
///
/// Deliberate carve-outs:
///  * positions with no usable tap (nil sample, char/alignment drift, or a
///    character with no key position) fall back to the static
///    `SpatialModel` cost — absent evidence must not change behavior,
///  * orthographic-confusion pairs (d→ð, t→þ, o→ö, ð↔þ) keep the static
///    confusion constant regardless of the tap: those are COGNITIVE slips —
///    the user deliberately hit the key they (wrongly) believed right, so
///    the tap point carries no information about the confusion,
///  * accent twins share their base key's center AND its personal
///    statistics (long-press callouts carry no tap sample, so the physical
///    key is the only touch target that trains), so their likelihood ratio
///    is 0 and the `minSubstitution` floor applies — identical to the
///    static model by construction.
///
/// `confidence(position:)` is the tap's NORMALIZED likelihood of the
/// resolved key over all physical keys (dead-center → ≈1, mid-boundary
/// between two keys → ≈0.5) under the same blended per-key distributions;
/// positions without a tap report 1.0 exactly (the static provider's
/// constant), so no-tap words aggregate to a neutral margin factor. The
/// corrector aggregates these into the autocorrect-margin veto and the
/// fold-pricing evidence penalty.
struct PerTapCostProvider: PositionCostProvider {

    /// Precomputed per-position evidence; nil where no usable tap exists.
    private struct Evidence {
        /// The resolved character the tap was recorded for (alignment guard).
        let char: Character
        /// Tap point in key-grid units (key center + normalized offsets).
        let point: SIMD2<Double>
        /// −log likelihood of the tap under the RESOLVED key's (blended)
        /// Gaussian, up to the shared prior constant.
        let nllResolved: Double
        /// Normalized likelihood of the resolved key at this tap.
        let confidence: Double
    }

    /// One key's blended Gaussian (stage 2): quadratic-form coefficients of
    /// the exponent ½·dᵀΣ⁻¹d around the mean-shifted center, plus the
    /// normalizer ½·ln|Σ| RELATIVE to the prior's (0 for the prior itself,
    /// so mixing personal and prior keys stays a coherent likelihood).
    private struct KeyGaussian {
        let center: SIMD2<Double>
        /// exponent = qxx·dx² + qxy·dx·dy + qyy·dy²  (d = point − center)
        let qxx: Double
        let qxy: Double
        let qyy: Double
        /// ½·ln(|Σ| / |Σ_prior|).
        let logWeight: Double
    }

    private let spatial: SpatialModel
    private let evidence: [Evidence?]
    private let invTwoSigmaX2: Double
    private let invTwoSigmaY2: Double
    private let minSubstitution: Double
    private let maxSubstitution: Double
    private let orthographicConfusion: Double
    /// Per-key personal Gaussians (keys past the min-samples gate, plus
    /// their accent twins); nil when no personal key is eligible — the
    /// stage-1 byte-identical path.
    private let personalGaussians: [Character: KeyGaussian]?

    /// Mean confidence over the positions that HAVE taps; nil when none
    /// do. The corrector's margin-veto aggregate.
    let meanTapConfidence: Double?

    /// Correlation clamp for the blended cross-covariance term (numerical
    /// safety: |ρ| → 1 makes Σ singular and the exponent explode along the
    /// anti-diagonal; genuine tap clouds measure well below this).
    private static let maxAbsCorrelation = 0.9

    /// - Parameters:
    ///   - taps: per typed-position samples, aligned with the (lowercased)
    ///     typed word; nil entries fall back to static pricing.
    ///   - personalTouch: the injected per-user touch snapshot (stage 2);
    ///     nil — or a snapshot with no key past the min-samples gate —
    ///     reproduces stage-1 pricing byte-identically.
    init(
        taps: [TapSample?],
        spatial: SpatialModel,
        config: EngineConfig,
        personalTouch: PersonalTouchSnapshot? = nil
    ) {
        self.spatial = spatial
        let costs = config.spatialCosts
        self.minSubstitution = costs.minSubstitution
        self.maxSubstitution = costs.maxSubstitution
        self.orthographicConfusion = costs.orthographicConfusion
        let ix = 1.0 / (2.0 * config.tapSigmaX * config.tapSigmaX)
        let iy = 1.0 / (2.0 * config.tapSigmaY * config.tapSigmaY)
        self.invTwoSigmaX2 = ix
        self.invTwoSigmaY2 = iy

        let personalGaussians = Self.blendGaussians(
            personalTouch: personalTouch, spatial: spatial, config: config)
        self.personalGaussians = personalGaussians

        /// −log likelihood of `point` under `key`'s effective Gaussian
        /// (personal-blend when eligible, else the shared prior), up to the
        /// prior's shared constant. Exactly the stage-1 arithmetic when the
        /// key has no personal Gaussian.
        func nll(from point: SIMD2<Double>, key: Character, center: SIMD2<Double>) -> Double {
            if let g = personalGaussians?[key] {
                let d = point - g.center
                return d.x * d.x * g.qxx + d.x * d.y * g.qxy + d.y * d.y * g.qyy + g.logWeight
            }
            let d = point - center
            return d.x * d.x * ix + d.y * d.y * iy
        }

        var evidence: [Evidence?] = []
        evidence.reserveCapacity(taps.count)
        var confidenceSum = 0.0
        var confidenceCount = 0
        for tap in taps {
            guard let tap, let center = spatial.keyCenter(of: tap.char) else {
                evidence.append(nil)
                continue
            }
            let point = center + SIMD2(tap.dxNorm, tap.dyNorm)
            let nllResolved = nll(from: point, key: tap.char, center: center)
            // Normalized likelihood of the resolved key over the physical
            // key set (accent twins share centers and are not separate
            // targets — summing them would double-count their base key).
            var total = 0.0
            for key in SpatialModel.physicalKeys {
                guard let keyCenter = spatial.keyCenter(of: key) else { continue }
                total += exp(-nll(from: point, key: key, center: keyCenter))
            }
            let confidence =
                total > 0 ? min(max(exp(-nllResolved) / total, 0), 1) : 1.0
            evidence.append(
                Evidence(
                    char: tap.char,
                    point: point,
                    nllResolved: nllResolved,
                    confidence: confidence
                )
            )
            confidenceSum += confidence
            confidenceCount += 1
        }
        self.evidence = evidence
        self.meanTapConfidence =
            confidenceCount > 0 ? confidenceSum / Double(confidenceCount) : nil
    }

    /// Build the per-key blended Gaussians (see the type doc's stage-2
    /// shrinkage). Returns nil when no key is eligible so the caller can
    /// keep the stage-1 arithmetic byte-identical. Per key, in order:
    /// min-samples gate → shrinkage blend → finiteness guard → σ floor →
    /// correlation clamp → determinant guard; any failure drops THAT key to
    /// the prior.
    private static func blendGaussians(
        personalTouch: PersonalTouchSnapshot?,
        spatial: SpatialModel,
        config: EngineConfig
    ) -> [Character: KeyGaussian]? {
        guard let personalTouch, !personalTouch.isEmpty else { return nil }
        let priorVarX = config.tapSigmaX * config.tapSigmaX
        let priorVarY = config.tapSigmaY * config.tapSigmaY
        let floorVar = config.touchSigmaFloor * config.touchSigmaFloor
        var gaussians: [Character: KeyGaussian] = [:]
        for key in SpatialModel.physicalKeys {
            guard
                let stats = personalTouch.stats(for: key),
                let meanOffset = personalTouch.blendedMeanOffset(for: key, config: config),
                let center = spatial.keyCenter(of: key),
                // Degenerate/corrupt raw aggregates (negative variance from
                // a mangled Welford M2, non-finite values) → this key stays
                // on the prior. `blendedMeanOffset` already vetoed
                // non-finite means/counts and the min-samples gate.
                stats.varianceX.isFinite, stats.varianceX >= 0,
                stats.varianceY.isFinite, stats.varianceY >= 0,
                stats.covarianceXY.isFinite
            else { continue }
            let weight = stats.count / (stats.count + config.touchPriorStrength)
            var varX = weight * stats.varianceX + (1 - weight) * priorVarX
            var varY = weight * stats.varianceY + (1 - weight) * priorVarY
            var cov = weight * stats.covarianceXY
            guard varX.isFinite, varY.isFinite, cov.isFinite, varX > 0, varY > 0 else {
                continue  // blend overflow paranoia — prior for this key
            }
            varX = max(varX, floorVar)
            varY = max(varY, floorVar)
            let maxCov = Self.maxAbsCorrelation * (varX * varY).squareRoot()
            cov = min(max(cov, -maxCov), maxCov)
            let det = varX * varY - cov * cov
            guard det > 0, det.isFinite else { continue }
            let gaussian = KeyGaussian(
                center: center + meanOffset,
                qxx: varY / (2 * det),
                qxy: -cov / det,
                qyy: varX / (2 * det),
                logWeight: 0.5 * log(det / (priorVarX * priorVarY))
            )
            gaussians[key] = gaussian
            // Accent twins share the physical key — and its personal cloud.
            for (accented, base) in SpatialModel.accentBase where base == key {
                gaussians[accented] = gaussian
            }
        }
        return gaussians.isEmpty ? nil : gaussians
    }

    /// −log likelihood of `point` under `intendedCenter`'s key Gaussian —
    /// the instance-side twin of the init-local `nll` (same arithmetic).
    private func nllIntended(
        point: SIMD2<Double>, intended: Character, intendedCenter: SIMD2<Double>
    ) -> Double {
        if let g = personalGaussians?[intended] {
            let d = point - g.center
            return d.x * d.x * g.qxx + d.x * d.y * g.qxy + d.y * d.y * g.qyy + g.logWeight
        }
        let d = point - intendedCenter
        return d.x * d.x * invTwoSigmaX2 + d.y * d.y * invTwoSigmaY2
    }

    func substitutionCost(position: Int, typed: Character, intended: Character) -> Double {
        if typed == intended { return 0 }
        // Cognitive, not spatial (see type doc): the static constant.
        if SpatialModel.confusionPairs.contains(String(typed) + String(intended)) {
            return orthographicConfusion
        }
        guard
            position >= 0, position < evidence.count,
            let evidence = evidence[position],
            evidence.char == typed,
            let intendedCenter = spatial.keyCenter(of: intended)
        else {
            return spatial.substitutionCost(typed: typed, intended: intended)
        }
        let nllIntended = nllIntended(
            point: evidence.point, intended: intended, intendedCenter: intendedCenter)
        // Log-likelihood ratio vs the resolved key ("0 when equal" holds by
        // construction), clamped into the static cost band so the beam's
        // caps/gates keep their calibrated meaning.
        let ratio = nllIntended - evidence.nllResolved
        return min(max(ratio, minSubstitution), maxSubstitution)
    }

    func confidence(position: Int) -> Double {
        guard position >= 0, position < evidence.count, let evidence = evidence[position] else {
            return 1.0
        }
        return evidence.confidence
    }

    /// Whether the position carries usable tap evidence (vs the static
    /// fallback). The margin veto only aggregates over these.
    func hasTap(at position: Int) -> Bool {
        position >= 0 && position < evidence.count && evidence[position] != nil
    }
}

extension PositionCostProvider {
    /// Evidence penalty a tap's confidence MULTIPLIES into the fold-pricing
    /// likelihood (−ln confidence in nats, capped): 0 for tapless positions
    /// and dead-center taps, up to `cap` for sloppy ones. See
    /// `FoldPricing`'s composition doc and `EngineConfig
    /// .tapFoldConfidenceMaxPenalty`.
    func foldEvidencePenalty(position: Int, cap: Double) -> Double {
        let confidence = self.confidence(position: position)
        guard confidence < 1 else { return 0 }
        return min(-log(max(confidence, 1e-9)), cap)
    }
}

File: AndroidClient/lib/engine/src/main/kotlin/is/solberg/lyklabord/engine/PerTapCostProvider.kt
package `is`.solberg.lyklabord.engine

import `is`.solberg.lyklabord.engine.config.EngineConfig
import kotlin.math.exp
import kotlin.math.ln
import kotlin.math.max
import kotlin.math.min
import kotlin.math.sqrt

/** One captured keystroke touch point in normalized key-pitch coordinates. */
data class TapSample(
    val char: Char,
    val dxNorm: Double,
    val dyNorm: Double,
)

private data class TapPoint(val x: Double, val y: Double) {
    operator fun plus(other: TapPoint): TapPoint = TapPoint(x + other.x, y + other.y)
    operator fun minus(other: TapPoint): TapPoint = TapPoint(x - other.x, y - other.y)
}

/** Coordinate-evidence position-cost provider used by the beam decoder. */
class PerTapCostProvider(
    taps: List<TapSample?>,
    private val spatial: SpatialModel,
    config: EngineConfig,
    personalTouch: PersonalTouchSnapshot? = null,
) : PositionCostProvider {
    private data class Evidence(
        val char: Char,
        val point: TapPoint,
        val center: TapPoint,
        val nllResolved: Double,
        val confidence: Double,
    )

    private data class KeyGaussian(
        val center: TapPoint,
        val qxx: Double,
        val qxy: Double,
        val qyy: Double,
        val logWeight: Double,
    )

    private val evidence: List<Evidence?>
    private val invTwoSigmaX2: Double
    private val invTwoSigmaY2: Double
    private val minSubstitution: Double
    private val maxSubstitution: Double
    private val orthographicConfusion: Double
    private val nearMissMinLean: Double
    private val nearMissCapEnabled: Boolean
    private val edgeUndershootEnabled: Boolean
    private val personalGaussians: Map<Char, KeyGaussian>?

    /** Mean confidence over positions carrying usable tap evidence. */
    val meanTapConfidence: Double?

    init {
        val costs = config.spatialCosts
        minSubstitution = costs.minSubstitution
        maxSubstitution = costs.maxSubstitution
        orthographicConfusion = costs.orthographicConfusion
        nearMissMinLean = config.tapNearMissMinLean
        nearMissCapEnabled = config.tapNearMissCapEnabled
        edgeUndershootEnabled = config.edgeUndershootEnabled
        invTwoSigmaX2 = 1.0 / (2.0 * config.tapSigmaX * config.tapSigmaX)
        invTwoSigmaY2 = 1.0 / (2.0 * config.tapSigmaY * config.tapSigmaY)

        personalGaussians = blendGaussians(personalTouch, spatial, config)

        fun nll(point: TapPoint, key: Char, center: TapPoint): Double {
            val gaussian = personalGaussians?.get(key)
            if (gaussian != null) {
                val delta = point - gaussian.center
                return delta.x * delta.x * gaussian.qxx +
                    delta.x * delta.y * gaussian.qxy +
                    delta.y * delta.y * gaussian.qyy + gaussian.logWeight
            }
            val delta = point - center
            return delta.x * delta.x * invTwoSigmaX2 + delta.y * delta.y * invTwoSigmaY2
        }

        val builtEvidence = ArrayList<Evidence?>(taps.size)
        var confidenceSum = 0.0
        var confidenceCount = 0
        for (tap in taps) {
            val centerRaw = tap?.let { spatial.keyCenter(it.char) }
            if (tap == null || centerRaw == null) {
                builtEvidence += null
                continue
            }
            val center = TapPoint(centerRaw.x, centerRaw.y)
            val point = center + TapPoint(tap.dxNorm, tap.dyNorm)
            val nllResolved = nll(point, tap.char, center)
            var total = 0.0
            for (key in SpatialModel.physicalKeys) {
                val keyCenterRaw = spatial.keyCenter(key) ?: continue
                val keyCenter = TapPoint(keyCenterRaw.x, keyCenterRaw.y)
                total += exp(-nll(point, key, keyCenter))
            }
            val confidence = if (total > 0.0) {
                min(max(exp(-nllResolved) / total, 0.0), 1.0)
            } else {
                1.0
            }
            builtEvidence += Evidence(tap.char, point, center, nllResolved, confidence)
            confidenceSum += confidence
            confidenceCount += 1
        }
        evidence = builtEvidence
        meanTapConfidence = if (confidenceCount > 0) {
            confidenceSum / confidenceCount.toDouble()
        } else {
            null
        }
    }

    private fun nllIntended(point: TapPoint, intended: Char, intendedCenter: TapPoint): Double {
        val gaussian = personalGaussians?.get(intended)
        if (gaussian != null) {
            val delta = point - gaussian.center
            return delta.x * delta.x * gaussian.qxx +
                delta.x * delta.y * gaussian.qxy +
                delta.y * delta.y * gaussian.qyy + gaussian.logWeight
        }
        val delta = point - intendedCenter
        return delta.x * delta.x * invTwoSigmaX2 + delta.y * delta.y * invTwoSigmaY2
    }

    override fun substitutionCost(position: Int, typed: Char, intended: Char): Double {
        if (typed == intended) return 0.0
        if (SpatialModel.confusionPairs.contains("$typed$intended")) {
            return orthographicConfusion
        }
        if (edgeUndershootEnabled &&
            SpatialModel.edgeUndershootPairs.contains("$typed$intended")
        ) {
            return spatial.substitutionCost(typed, intended)
        }
        if (position !in evidence.indices) {
            return spatial.substitutionCost(typed, intended)
        }
        val currentEvidence = evidence[position]
            ?: return spatial.substitutionCost(typed, intended)
        if (currentEvidence.char != typed) {
            return spatial.substitutionCost(typed, intended)
        }
        val intendedCenterRaw = spatial.keyCenter(intended)
            ?: return spatial.substitutionCost(typed, intended)
        val intendedCenter = TapPoint(intendedCenterRaw.x, intendedCenterRaw.y)
        val intendedNll = nllIntended(currentEvidence.point, intended, intendedCenter)
        val ratio = intendedNll - currentEvidence.nllResolved
        val clamped = min(max(ratio, minSubstitution), maxSubstitution)
        if (nearMissCapEnabled && clamped > minSubstitution &&
            tapLean(currentEvidence, intendedCenter) >= nearMissMinLean
        ) {
            return min(clamped, spatial.substitutionCost(typed, intended))
        }
        return clamped
    }

    private fun tapLean(evidence: Evidence, towards: TapPoint): Double {
        val direction = towards - evidence.center
        val length = sqrt(direction.x * direction.x + direction.y * direction.y)
        if (length <= 0.0) return 0.0
        val offset = evidence.point - evidence.center
        return (offset.x * direction.x + offset.y * direction.y) / length
    }

    /** Whether this tap supports reading [typed] as [intended]. */
    fun tapSupports(position: Int, typed: Char, intended: Char): Boolean {
        if (typed == intended) return false
        if (edgeUndershootEnabled &&
            SpatialModel.edgeUndershootPairs.contains("$typed$intended")
        ) {
            return true
        }
        if (!nearMissCapEnabled || position !in evidence.indices) return false
        val currentEvidence = evidence[position] ?: return false
        if (currentEvidence.char != typed) return false
        val intendedCenterRaw = spatial.keyCenter(intended) ?: return false
        return tapLean(currentEvidence, TapPoint(intendedCenterRaw.x, intendedCenterRaw.y)) >=
            nearMissMinLean
    }

    override fun confidence(position: Int): Double =
        if (position !in evidence.indices) 1.0 else evidence[position]?.confidence ?: 1.0

    /** Whether the position carries usable tap evidence rather than static fallback. */
    override fun hasTap(position: Int): Boolean =
        position in evidence.indices && evidence[position] != null

    private companion object {
        const val MAX_ABS_CORRELATION = 0.9

        fun blendGaussians(
            personalTouch: PersonalTouchSnapshot?,
            spatial: SpatialModel,
            config: EngineConfig,
        ): Map<Char, KeyGaussian>? {
            if (personalTouch == null || personalTouch.isEmpty) return null
            val priorVarX = config.tapSigmaX * config.tapSigmaX
            val priorVarY = config.tapSigmaY * config.tapSigmaY
            val floorVar = config.touchSigmaFloor * config.touchSigmaFloor
            val gaussians = HashMap<Char, KeyGaussian>()
            for (key in SpatialModel.physicalKeys) {
                val stats = personalTouch.stats(key) ?: continue
                val meanOffset = personalTouch.blendedMeanOffset(key, config) ?: continue
                val centerRaw = spatial.keyCenter(key) ?: continue
                if (!stats.varianceX.isFinite() || stats.varianceX < 0.0 ||
                    !stats.varianceY.isFinite() || stats.varianceY < 0.0 ||
                    !stats.covarianceXY.isFinite()
                ) continue
                val weight = stats.count / (stats.count + config.touchPriorStrength)
                var varX = weight * stats.varianceX + (1.0 - weight) * priorVarX
                var varY = weight * stats.varianceY + (1.0 - weight) * priorVarY
                var cov = weight * stats.covarianceXY
                if (!varX.isFinite() || !varY.isFinite() || !cov.isFinite() ||
                    varX <= 0.0 || varY <= 0.0
                ) continue
                varX = max(varX, floorVar)
                varY = max(varY, floorVar)
                val maxCov = MAX_ABS_CORRELATION * sqrt(varX * varY)
                cov = min(max(cov, -maxCov), maxCov)
                val determinant = varX * varY - cov * cov
                if (determinant <= 0.0 || !determinant.isFinite()) continue
                val gaussian = KeyGaussian(
                    center = TapPoint(
                        centerRaw.x + meanOffset.first,
                        centerRaw.y + meanOffset.second,
                    ),
                    qxx = varY / (2.0 * determinant),
                    qxy = -cov / determinant,
                    qyy = varX / (2.0 * determinant),
                    logWeight = 0.5 * ln(determinant / (priorVarX * priorVarY)),
                )
                gaussians[key] = gaussian
                for ((accented, base) in SpatialModel.accentBase) {
                    if (base == key) gaussians[accented] = gaussian
                }
            }
            return gaussians.ifEmpty { null }
        }
    }
}

/** Evidence penalty applied by fold pricing to a position's tap confidence. */
fun PositionCostProvider.foldEvidencePenalty(position: Int, cap: Double): Double {
    val confidence = confidence(position)
    if (confidence >= 1.0) return 0.0
    return min(-ln(max(confidence, 1e-9)), cap)
}
